platforms/core-runtime/build-operations/src/test/groovy/org/gradle/internal/operations/DefaultBuildOperationRunnerTest.groovy

```diff
@@ -252,6 +254,474 @@ class DefaultBuildOperationRunnerTest extends Specification {
         })
     }
 
+    def "action can mark operation as failed without throwing an exception"() {
+        setup:
+        def buildOperation = Spy(TestRunnableBuildOperation)
+        def failure = new RuntimeException()
+
+        when:
+        operationRunner.run(buildOperation)
+
+        then:
+        1 * buildOperation.run(_) >> { BuildOperationContext context -> context.failed(failure) }
+
+        then:
+        1 * listener.stop(_, _, _, _) >> { BuildOperationDescriptor descriptor, BuildOperationState operationState, @Nullable BuildOperationState parent, ReadableBuildOperationContext context ->
+            assert context.failure == failure
+        }
+    }
```

Comment:
💭 These tests were airlifted from `DefaultBuildOperationExecutorTest` that was not, in fact, testing a `BuildOperationExecutor`, but a runner.